projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/util/LLVMControlFlowGraph.java

```diff
@@ -0,0 +1,335 @@
+
+    private int nextLoop = 0;
+
+    private boolean reducible = true;
+
+    public final class CFGBlock {
+
+        private final InstructionBlock instructionBlock;
+
+        public int id;
+        public List<CFGBlock> sucs = new ArrayList<>();
+        public List<CFGBlock> preds = new ArrayList<>();
+
+        public boolean visited = false;
+        public boolean active = false;
+        public boolean isLoopHeader = false;
+        public long loops;  // could use bitmap instead
+        public int loopId;
+
+        public CFGBlock(InstructionBlock block) {
+            this.instructionBlock = block;
+            this.id = block.getBlockIndex();
+        }
+
+        @Override
+        public String toString() {
+            return instructionBlock.toString();
+        }
+
+    }
+
+    public final class CFGLoop {
+        private CFGBlock loopHeader;
+        private List<CFGBlock> body;
+        private List<CFGLoop> innerLoops;
+
+        private Set<CFGBlock> successors;
+
+        private final int id;
+
+        public CFGLoop(int id) {
+            this.id = id;
+            body = new ArrayList<>();
+            innerLoops = new ArrayList<>();
+        }
+
+        public List<CFGBlock> getBody() {
+            return body;
+        }
+
+        public Set<CFGBlock> getSuccessors() {
+            if (successors == null) {
+                calculateSuccessors();
+            }
+
+            return this.successors;
+        }
+
+        /**
+         * Calculates the successors of this loop with successors of inner loops potentially forwarded to
+         * the outer loop
+         */
+        private void calculateSuccessors() {
+            successors = new HashSet<>();
+
+            for (CFGBlock s : loopHeader.sucs) {
+                if (!isInLoop(s)) {
+                    successors.add(s);
+                }
+            }
+
+            for (CFGBlock b : body) {
+                // for each inner loop, add all successors which are not in the outer loop
+                // to the successors of the outer loop
+                if (b.isLoopHeader) {
+                    for (CFGLoop l : innerLoops) {
+                        if (l.getHeader().equals(b)) {
+                            for (CFGBlock ib : l.getSuccessors()) {
+                                if (!isInLoop(ib)) {
+                                    successors.add(ib);
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    for (CFGBlock s : b.sucs) {
+                        if (!isInLoop(s)) {
+                            successors.add(s);
+                        }
+                    }
+                }
+            }
+
+        }
+
+        public Integer[] getSuccessorIDs() {
+            if (successors == null) {
+                calculateSuccessors();
+            }
+
+            Integer[] sucIDs = new Integer[successors.size()];
```

Comment:
Is the order of elements in `sucIDs` guaranteed by your loop detection algorithm? These seem like they should be sorted in some way.